core/src/main/java/com/segment/analytics/kotlin/core/Telemetry.kt

```diff
@@ -0,0 +1,272 @@
+        if (tags.isEmpty()) return
+        if (Math.random() > sampleRate) return
+        if (queue.size >= maxQueueSize) return
+
+        addRemoteMetric(metric, tags, value = (1.0 / sampleRate).roundToInt())
+    }
+
+    fun error(metric:String, tags: Map<String, String>, log: String) {
+        if (!enable) return
+        if (!metric.startsWith("analytics_mobile.")) return
+        if (tags.isEmpty()) return
+        if (queue.size >= maxQueueSize) return
+
+        var filteredTags = tags
+        if (!sendWriteKeyOnError) filteredTags = tags.filterKeys { it != "writekey" }
+        var logData: String? = null
+        if (sendErrorLogData) logData = log
+
+        val errorKey = tags["error"]
+        if (errorKey != null) {
+            if (seenErrors.containsKey(errorKey)) {
+                seenErrors[errorKey] = seenErrors[errorKey]!! + 1
+                if (Math.random() > sampleRate) return
+                // Send how many we've seen after the first since we know for sure.
+                addRemoteMetric(metric, filteredTags, log=logData, value = seenErrors[errorKey]!!)
+                seenErrors[errorKey] = 0
+            } else {
+                addRemoteMetric(metric, filteredTags, log=logData)
+                flush()
+                seenErrors[errorKey] = 0 // Zero because it's already been sent.
+            }
+        }
+    }
+
+    fun flush() {
+        if (!started || !enable || queue.isEmpty()) return
+
+        if (_rateLimitTimer > 0 && _rateLimitTimer > (System.currentTimeMillis() / 1000).toInt()) {
+            return
+        }
+        _rateLimitTimer = 0
+
+        try {
+            send()
+            queueBytes = 0
+        } catch (error: Throwable) {
+            logError(error)
+            sampleRate = 0.0
+        }
+    }
+
+    private fun send() {
+        val payload = Json.encodeToString(mapOf("series" to queue))
+        resetQueue()
+
+        try {
+            val connection = httpClient.upload(host)
+            connection.outputStream?.use { outputStream ->
+                // Write the JSON string to the outputStream.
+                outputStream.write(payload.toByteArray(Charsets.UTF_8))
+                outputStream.flush() // Ensure all data is written
+            }
+            connection.close()
+        } catch (e: HTTPException) {
+            _errorHandler?.let { it(e) }
+            if (e.responseCode != 429) {
+                val headers = e.responseHeaders
+                val rateLimit = headers["X-RateLimit-Reset"]?.firstOrNull()?.toLongOrNull()
+                if (rateLimit != null) {
+                    _rateLimitTimer = rateLimit.toInt() + (System.currentTimeMillis() / 1000).toInt()
+                }
+            }
+        } catch (e: Exception) {
+            _errorHandler?.let { it(e) }
+        }
+    }
+
+    private val additionalTags: Map<String, String> by lazy {
+        var osVersion = System.getProperty("os.version")
```

Comment:
There is probably already code to get this information somewhere?

Reply:
There is in the android context plugin.